js/test/wallet/createTransaction.test.ts

```diff
@@ -0,0 +1,194 @@
+                    );
+                }).rejects.toThrow('payer owner invalid');
+            });
+
+            it.todo('throws an error on executable payer');
+
+            it('throws an error on invalid recipient owner', async () => {
+                expect.assertions(1);
+
+                await expect(async () => {
+                    return await createTransaction(
+                        connection,
+                        new PublicKey('mvines9iiHiQTysrwkJjGf2gb9Ex9jXJX8ns3qwf2kN'),
+                        new PublicKey(SystemProgram.programId),
+                        new BigNumber(1),
+                        {}
+                    );
+                }).rejects.toThrow('recipient owner invalid');
+            });
+
+            it.todo('throws an error on executable recipient');
+
+            it('throws an error on invalid decimal amount', async () => {
+                expect.assertions(1);
+
+                await expect(async () => {
+                    return await createTransaction(
+                        connection,
+                        wallet.publicKey,
+                        new PublicKey('mvines9iiHiQTysrwkJjGf2gb9Ex9jXJX8ns3qwf2kN'),
+                        new BigNumber(1.0000000000001),
+                        {}
+                    );
+                }).rejects.toThrow('amount decimals invalid');
+            });
+
+            it('throws an error on insufficient funds', async () => {
+                expect.assertions(1);
+
+                await expect(async () => {
+                    return await createTransaction(
+                        connection,
+                        wallet.publicKey,
+                        new PublicKey('mvines9iiHiQTysrwkJjGf2gb9Ex9jXJX8ns3qwf2kN'),
+                        new BigNumber(100),
+                        {}
+                    );
+                }).rejects.toThrow('insufficient funds');
+            });
+        });
+    });
+
+    describe('when transferring SPL token', () => {
+        describe('errors', () => {
+            it.todo('throws an error on invalid mint');
+
+            it('throws an error on invalid decimal amount', async () => {
+                expect.assertions(1);
+
+                await expect(async () => {
+                    return await createTransaction(
+                        connection,
+                        wallet.publicKey,
+                        new PublicKey('mvines9iiHiQTysrwkJjGf2gb9Ex9jXJX8ns3qwf2kN'),
+                        new BigNumber(1.0000000001),
+                        {
+                            token: new PublicKey('8FRFC6MoGGkMFQwngccyu69VnYbzykGeez7ignHVAFSN'),
+                        }
+                    );
+                }).rejects.toThrow('amount decimals invalid');
+            });
+
+            // FIXME: can't get past this error
+            // Fails at getting associated token account
+            it.skip('throws an error on unintialized payer', async () => {
```

Comment:
@jordansexton for the implementation, we may need to consider having a getOrCreateAssociatedTokenAccount

Reply:
No, the payer and recipient need to exist first, we don't need to create them. It's up to wallets to ensure they do (but you can't send USDC unless you already have the ATA, so this is moot). For testing purposes, we can just require that you run the test validator and create some accounts on localnet in the setup I imagine?

Reply:
hmm, I guess the flow would be for SPL:
1. Wallet parses the URL
2. Gets the token and creates the ATA if it doesn't exist
3. Then creates the transaction

Reply:
No, not quite. 2 is what doesn't make sense -- a user can't send, say, USDC, unless they already have an ATA for USDC. A wallet doesn't need to create an ATA for you (you either have it because you have the tokens or you don't).

Reply:
ahh sorry yes! ofcourse.